File: frontend/types/dataSource.ts
export type SourceType = "FILE" | "WEB";
export type DataSourceStatus = "PREPARING" | "TRAINING" | "AVAILABLE" | "ERROR";
export type Priority = "HIGH" | "MEDIUM" | "LOW";
export type SortColumn = "id" | "title" | "updated_at";
export type SortOrder = "asc" | "desc";

export type DataSourceClassification = {
  type_code: string;
  classification_type_id: number;
  classification_value_id: number;
  display_label: string;
  value_name: string;
};

export type DataSource = {
  id: number;
  source_type: SourceType;
  title: string;
  format: string;
  status: DataSourceStatus;
  category_name: string | null;
  size_bytes: number | null;
  character_count: number | null;
  answer_source_enabled: boolean;
  priority: Priority;
  reference_link_visible: boolean;
  updated_at: string;
  version: number;
  file: { file_name: string } | null;
  website: { url: string } | null;
  classifications: DataSourceClassification[];
};

export type DataSourceFilters = {
  keyword: string;
  format: string;
  status: string;
  type_1_value_id: string;
  type_2_value_id: string;
  type_3_value_id: string;
  answer_source_enabled: string;
  priority: string;
  reference_link_visible: string;
  sort: SortColumn;
  order: SortOrder;
  page: number;
  page_size: number;
};

export type DataSourceListResponse = {
  items: DataSource[];
  page: number;
  page_size: number;
  total_count: number;
  total_pages: number;
  total_size_bytes: number;
  sort: SortColumn;
  order: SortOrder;
};

export type FileDataSourceUpdate = {
  title: string;
  type_1_value_id: number | null;
  type_2_value_id: number | null;
  type_3_value_id: number | null;
  priority: Priority;
  answer_source_enabled: boolean;
  reference_link_visible: boolean;
  version: number;
};

export type WebsiteDataSourceCreate = {
  url: string;
  title: string;
  type_1_value_id: number | null;
  type_2_value_id: number | null;
  type_3_value_id: number | null;
  priority: Priority;
  answer_source_enabled: boolean;
  reference_link_visible: boolean;
};

export class DataSourcesApiError extends Error {
  constructor(message: string, public status: number, public code?: string) {
    super(message);
  }
}

File: backend/app/schemas/data_source.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


SourceType = Literal["FILE", "WEB"]
DataSourceStatus = Literal["PREPARING", "TRAINING", "AVAILABLE", "ERROR"]
Priority = Literal["HIGH", "MEDIUM", "LOW"]
SortColumn = Literal["id", "title", "updated_at"]
SortOrder = Literal["asc", "desc"]


class DataSourceFileResponse(BaseModel):
    file_name: str


class DataSourceWebsiteResponse(BaseModel):
    url: str
    last_fetched_at: datetime | None


class DataSourceClassificationResponse(BaseModel):
    type_code: str
    classification_type_id: int
    classification_value_id: int
    display_label: str
    value_name: str


class DataSourceCategoryResponse(BaseModel):
    id: int
    name: str
    parent_id: int | None
    path: str


class DataSourceResponse(BaseModel):
    id: int
    source_type: SourceType
    title: str
    format: str
    status: DataSourceStatus
    category: DataSourceCategoryResponse | None
    category_name: str | None
    size_bytes: int | None
    character_count: int | None
    answer_source_enabled: bool
    priority: Priority
    reference_link_visible: bool
    updated_at: datetime
    version: int
    file: DataSourceFileResponse | None
    website: DataSourceWebsiteResponse | None
    classifications: list[DataSourceClassificationResponse]


class DataSourceListResponse(BaseModel):
    items: list[DataSourceResponse]
    page: int
    page_size: int
    total_count: int
    total_pages: int
    total_size_bytes: int
    sort: SortColumn
    order: SortOrder


class DataSourceFilters(BaseModel):
    keyword: str | None = None
    format: str | None = None
    status: DataSourceStatus | None = None
    category_id: int | None = Field(default=None, ge=1)
    type_1_value_id: int | None = None
    type_2_value_id: int | None = None
    type_3_value_id: int | None = None
    answer_source_enabled: bool | None = None
    priority: Priority | None = None
    reference_link_visible: bool | None = None
    sort: SortColumn = "updated_at"
    order: SortOrder = "desc"
    page: int = Field(default=1, ge=1)
    page_size: int = 10

    @field_validator("keyword")
    @classmethod
    def normalize_keyword(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None

    @field_validator("page_size")
    @classmethod
    def validate_page_size(cls, value: int) -> int:
        if value not in (10, 20, 50, 100):
            raise ValueError("表示件数は10、20、50、100のいずれかを指定してください。")
        return value


class ToggleAnswerSourceRequest(BaseModel):
    enabled: bool
    version: int = Field(ge=1)


class ToggleReferenceLinkRequest(BaseModel):
    visible: bool
    version: int = Field(ge=1)


class DeleteTarget(BaseModel):
    id: int
    version: int = Field(ge=1)


class BulkDeleteRequest(BaseModel):
    items: list[DeleteTarget] = Field(min_length=1)


class BulkDeleteResponse(BaseModel):
    deleted_count: int


class ClassificationAssignment(BaseModel):
    classification_type_id: int
    classification_value_id: int


class FileUploadResponse(BaseModel):
    items: list[DataSourceResponse]
    created_count: int


class FileDataSourceUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str = Field(max_length=500)
    category_id: int | None = Field(default=None, ge=1)
    type_1_value_id: int | None = Field(default=None, ge=1)
    type_2_value_id: int | None = Field(default=None, ge=1)
    type_3_value_id: int | None = Field(default=None, ge=1)
    priority: Priority
    answer_source_enabled: bool
    reference_link_visible: bool
    version: int = Field(ge=1)


class WebsiteDataSourceCreateRequest(BaseModel):
    url: str = ""
    title: str = ""
    category_id: int | None = Field(default=None, ge=1)
    type_1_value_id: int | None = Field(default=None, ge=1)
    type_2_value_id: int | None = Field(default=None, ge=1)
    type_3_value_id: int | None = Field(default=None, ge=1)
    priority: str = "MEDIUM"
    answer_source_enabled: bool = True
    reference_link_visible: bool = True


class WebsiteDataSourceUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    url: str
    title: str
    category_id: int | None = Field(default=None, ge=1)
    type_1_value_id: int | None = Field(default=None, ge=1)
    type_2_value_id: int | None = Field(default=None, ge=1)
    type_3_value_id: int | None = Field(default=None, ge=1)
    priority: str
    answer_source_enabled: bool
    reference_link_visible: bool
    version: int = Field(ge=1)
